app/react/src/server/framework-preset-react.ts

```diff
@@ -20,13 +21,28 @@ export async function babel(config: TransformOptions, options: StorybookOptions)
     plugins: [require.resolve('react-refresh/babel'), ...(config.plugins || [])],
   };
 }
+const storybookReactDirName = path.dirname(require.resolve('@storybook/react/package.json'));
+// TODO: improve node_modules detection
+const context = storybookReactDirName.includes('node_modules')
+  ? path.join(storybookReactDirName, '../../') // Real life case, already in node_modules
+  : path.join(storybookReactDirName, '../../node_modules'); // SB Monorepo
+
+const hasJsxRuntime = () => {
+  try {
+    require.resolve('react/jsx-runtime', { paths: [context] });
+    return true;
+  } catch (e) {
+    return false;
+  }
+};
 
 export async function babelDefault(config: TransformOptions) {
+  const presetReactOptions = hasJsxRuntime() ? { runtime: 'automatic' } : {};
```

Comment:
Wouldn't a better approach be to use `path.dirname(config.filename)` to get the directory for the file being transformed, pass that to `hasJsxRuntime`, and use it as the `paths` value for the `require.resolve` lookup (rather than poking into an arbitrary nm folder)?

Reply:
Definitely ! We have this behavior in other places that needs to be replaced. 
I'm interested in creating a @storybook/paths internal package or similar stuff to unify how Storybook infers with client's setup. We had a brief chat with @mrmckeb @shilman @yannbf about it :)